File: compiler/engine/quantizer.py
"""
Shannon Quantizer
Post-Training Symmetric INT8 / INT4 Quantization Engine with Mixed-Precision Support.
"""

from typing import Tuple, Dict, Any, List
import numpy as np
from .ir import ModelGraph, Tensor, Layer

class Quantizer:
    def __init__(self, bits: int = 8, symmetric: bool = True, mixed_precision: bool = False):
        self.bits = bits
        self.symmetric = symmetric
        self.mixed_precision = mixed_precision
        self.qmin = -(2 ** (bits - 1)) if symmetric else 0
        self.qmax = (2 ** (bits - 1)) - 1 if symmetric else (2 ** bits) - 1

    def _get_layer_bitwidth(self, layer_idx: int, total_layers: int, op_type: str) -> int:
        """
        Mixed-Precision policy (based on HAWQ research):
        - First Conv and Final Classifier layers: High sensitivity -> INT8
        - Intermediate deep projection layers: Low sensitivity -> INT4
        """
        if not self.mixed_precision:
            return self.bits
        if layer_idx == 0 or layer_idx == total_layers - 1:
            return 8
        if op_type in ["Dense", "Conv2D"]:
            return 4
        return 8

    def quantize_tensor(self, tensor: Tensor, target_bits: int = 8) -> Tuple[Tensor, float, int]:
        if tensor.data is None or tensor.dtype in ["int8", "int4"]:
            return tensor, tensor.scale, tensor.zero_point

        data = tensor.data.astype(np.float32)
        min_val = float(np.min(data))
        max_val = float(np.max(data))

        qmin = -(2 ** (target_bits - 1)) if self.symmetric else 0
        qmax = (2 ** (target_bits - 1)) - 1 if self.symmetric else (2 ** target_bits) - 1

        if self.symmetric:
            max_abs = max(abs(min_val), abs(max_val), 1e-7)
            scale = max_abs / qmax
            zero_point = 0
            qdata = np.clip(np.round(data / scale), qmin, qmax).astype(np.int8)
        else:
            scale = max((max_val - min_val) / (qmax - qmin), 1e-7)
            zero_point = int(np.clip(np.round(-min_val / scale), qmin, qmax))
            qdata = np.clip(np.round(data / scale) + zero_point, qmin, qmax).astype(np.int8)

        quantized_tensor = Tensor(
            name=tensor.name,
            shape=tensor.shape,
            dtype="int8" if target_bits == 8 else "int4",
            data=qdata
        )
        quantized_tensor.scale = scale
        quantized_tensor.zero_point = zero_point
        return quantized_tensor, scale, zero_point

    def quantize_graph(self, graph: ModelGraph) -> ModelGraph:
        total_layers = len(graph.layers)
        for idx, layer in enumerate(graph.layers):
            layer_bits = self._get_layer_bitwidth(idx, total_layers, layer.op_type)
            if layer.weights is not None:
                layer.weights, _, _ = self.quantize_tensor(layer.weights, target_bits=layer_bits)
            if layer.bias is not None:
                layer.bias.dtype = "int32"
                layer.bias.size_bytes = int(np.prod(layer.bias.shape)) * 4 if layer.bias.shape else 4
                if layer.bias.data is not None:
                    w_scale = layer.weights.scale if layer.weights else 1.0
                    b_scale = w_scale * 0.01
                    layer.bias.data = np.clip(np.round(layer.bias.data / b_scale), -2147483648, 2147483647).astype(np.int32)
                    layer.bias.scale = b_scale

        for tensor_name, tensor in graph.tensors.items():
            if tensor.dtype == "float32":
                tensor.dtype = "int8"
                tensor.size_bytes = int(np.prod(tensor.shape)) if tensor.shape else 1

        graph.compute_stats()
        return graph

File: frontend/src/compiler/quantizer.ts
import { ModelGraph, Tensor, Layer } from './ir';
import { QuantizationMetrics, LayerQuantMetric } from '../types';

export interface QuantizerConfig {
  bits: 4 | 8;
  symmetric: boolean;
  mixed_precision: boolean;
}

export class Quantizer {
  config: QuantizerConfig;

  constructor(config: Partial<QuantizerConfig> = {}) {
    this.config = {
      bits: config.bits || 8,
      symmetric: config.symmetric !== undefined ? config.symmetric : true,
      mixed_precision: config.mixed_precision || false,
    };
  }

  /**
   * Quantizes an arbitrary Float32Array tensor into symmetric signed integer values (INT8 / INT4).
   */
  quantizeTensor(tensor: Tensor, targetBits: 4 | 8 = 8): Tensor {
    if (!tensor.data) {
      const qTensor = new Tensor(tensor.name, tensor.shape, `int${targetBits}`);
      qTensor.scale = 0.0078125;
      qTensor.zero_point = 0;
      return qTensor;
    }

    const data = tensor.data;
    const len = data.length;
    let maxVal = 0;

    for (let i = 0; i < len; i++) {
      const abs = Math.abs(data[i]);
      if (abs > maxVal) maxVal = abs;
    }

    maxVal = Math.max(maxVal, 1e-7);

    // Symmetric scale factor: S = max(|w|) / (2^(b-1) - 1)
    const qMax = (1 << (targetBits - 1)) - 1;
    const qMin = -(1 << (targetBits - 1));
    const scale = maxVal / qMax;

    const quantized = new Int8Array(len);
    for (let i = 0; i < len; i++) {
      const q = Math.round(data[i] / scale);
      quantized[i] = Math.max(qMin, Math.min(qMax, q));
    }

    const qTensor = new Tensor(tensor.name, tensor.shape, `int${targetBits}`, quantized);
    qTensor.scale = scale;
    qTensor.zero_point = 0;
    return qTensor;
  }

  /**
   * Clones and quantizes an entire ModelGraph IR, applying symmetric quantization to all weight tensors.
   */
  quantizeGraph(graph: ModelGraph): ModelGraph {
    const qGraph = new ModelGraph(graph.name);

    for (const [, tensor] of Object.entries(graph.tensors)) {
      qGraph.addTensor(new Tensor(tensor.name, tensor.shape, tensor.dtype));
    }

    for (const layer of graph.layers) {
      const qLayer = new Layer(
        layer.layer_id,
        layer.op_type,
        [...layer.inputs],
        [...layer.outputs],
        { ...layer.params }
      );
      let targetBits = this.config.bits;

      if (this.config.mixed_precision && (layer.op_type === 'Dense' || layer.op_type === 'FullyConnected')) {
        targetBits = 4;
      }

      if (layer.weights) {
        qLayer.weights = this.quantizeTensor(layer.weights, targetBits);
      }

      if (layer.bias) {
        qLayer.bias = this.quantizeTensor(layer.bias, 8);
      }

      qGraph.addLayer(qLayer);
    }

    qGraph.inputs = [...graph.inputs];
    qGraph.outputs = [...graph.outputs];
    return qGraph;
  }
}

/**
 * Computes exact mathematical quantization error telemetry:
 * Signal-to-Quantization-Noise Ratio (SQNR in dB), Mean Squared Error (MSE),
 * exact Cosine Similarity, and peak absolute numerical delta.
 */
export function computeQuantizationMetrics(
  originalGraph: ModelGraph,
  quantizedGraph: ModelGraph
): QuantizationMetrics {
  const layerMetrics: LayerQuantMetric[] = [];
  let totalSignalPower = 0;
  let totalNoisePower = 0;
  let totalElements = 0;
  let overallMaxError = 0;
  let totalDotProd = 0;
  let totalNormOrig = 0;
  let totalNormQuant = 0;

  for (let idx = 0; idx < originalGraph.layers.length; idx++) {
    const origLayer = originalGraph.layers[idx];
    const quantLayer = quantizedGraph.layers[idx];

    if (!quantLayer) continue;

    const origW = origLayer.weights;
    const quantW = quantLayer.weights;

    if (origW && quantW && origW.data && quantW.data && origW.data.length > 0) {
      const len = origW.data.length;
      let layerSigPower = 0;
      let layerNoisePower = 0;
      let layerMaxErr = 0;
      let layerDot = 0;
      let layerNormOrig = 0;
      let layerNormQuant = 0;
      const scale = quantW.scale || 1.0;

      const sampleSize = Math.min(len, 64);
      const step = Math.max(1, Math.floor(len / sampleSize));
      const sampleFp32: number[] = [];
      const sampleInt8: number[] = [];

      for (let i = 0; i < len; i++) {
        const fpVal = origW.data[i];
        const intVal = quantW.data[i];
        const dequantVal = intVal * scale;
        const err = fpVal - dequantVal;
        const absErr = Math.abs(err);

        layerSigPower += fpVal * fpVal;
        layerNoisePower += err * err;
        layerDot += fpVal * dequantVal;
        layerNormOrig += fpVal * fpVal;
        layerNormQuant += dequantVal * dequantVal;

        if (absErr > layerMaxErr) layerMaxErr = absErr;

        if (i % step === 0 && sampleFp32.length < sampleSize) {
          sampleFp32.push(Number(fpVal.toFixed(4)));
          sampleInt8.push(Number(dequantVal.toFixed(4)));
        }
      }

      const layerMse = layerNoisePower / len;
      const layerSqnr = 10 * Math.log10(Math.max(1e-12, layerSigPower) / Math.max(1e-12, layerNoisePower));
      const denom = Math.sqrt(layerNormOrig) * Math.sqrt(layerNormQuant);
      const layerCosSim = denom > 0 ? layerDot / denom : 1.0;

      totalSignalPower += layerSigPower;
      totalNoisePower += layerNoisePower;
      totalDotProd += layerDot;
      totalNormOrig += layerNormOrig;
      totalNormQuant += layerNormQuant;
      totalElements += len;

      if (layerMaxErr > overallMaxError) overallMaxError = layerMaxErr;

      layerMetrics.push({
        layer_id: origLayer.layer_id,
        mse: Number(layerMse.toFixed(6)),
        sqnr_db: Number(layerSqnr.toFixed(2)),
        max_error: Number(layerMaxErr.toFixed(6)),
        cosine_similarity: Number(layerCosSim.toFixed(5)),
        sample_fp32: sampleFp32,
        sample_int8: sampleInt8,
      });
    } else {
      layerMetrics.push({
        layer_id: origLayer.layer_id,
        mse: 0,
        sqnr_db: 99.9,
        max_error: 0,
        cosine_similarity: 1.0,
        sample_fp32: [],
        sample_int8: [],
      });
    }
  }

  const globalMse = totalElements > 0 ? totalNoisePower / totalElements : 0.0001;
  const globalSqnr = totalElements > 0
    ? 10 * Math.log10(Math.max(1e-12, totalSignalPower) / Math.max(1e-12, totalNoisePower))
    : 48.0;

  const globalDenom = Math.sqrt(totalNormOrig) * Math.sqrt(totalNormQuant);
  const globalCosSim = globalDenom > 0 ? totalDotProd / globalDenom : 0.9998;

  return {
    sqnr_db: Number(globalSqnr.toFixed(2)),
    mse: Number(globalMse.toFixed(6)),
    max_error: Number(overallMaxError.toFixed(6)),
    cosine_similarity: Number(globalCosSim.toFixed(5)),
    layer_metrics: layerMetrics,
  };
}
